### PR TITLE
WIP: feature: introduce ModelBundle

### PR DESCRIPTION
**Pull request by mxmCherry**
This would allow to cache all the models to make this lib service-ready
(init once, analyze multiple audio-s).

### CODE REVIEW

File: pyAudioAnalysis/modelBundle.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import pickle
+
+class ModelBundle:
```
Comment (dim) [inline pyAudioAnalysis/modelBundle.py:3]: So I am not a fan of this. Why not simply 3 separate HMMModel, SVMModel and KNNModel classes?

File: pyAudioAnalysis/audioSegmentation2.py [hunk 2]
```diff
@@ -466,27 +467,25 @@ def save_hmm(hmm_model_name, model, classes, mid_window, mid_step):
 
 
 def hmm_segmentation(audio_file, hmm_model_name, plot_results=False,
```
Comment (dim) [inline pyAudioAnalysis/audioSegmentation2.py:469]: I suggest you change this to: `def hmm_segmentation(audio_file, hmm_model, plot_results=False, ground_truth="")` and ...

File: pyAudioAnalysis/audioSegmentation2.py [hunk 3]
```diff
@@ -466,27 +467,25 @@ def save_hmm(hmm_model_name, model, classes, mid_window, mid_step):
...
-                     gt_file=""):
+                     gt_file="", models=None):
     sampling_rate, signal = audioBasicIO.read_audio_file(audio_file)
 
-    with open(hmm_model_name, "rb") as f_handle:
-        hmm = cpickle.load(f_handle)
-        class_names = cpickle.load(f_handle)
-        mid_window = cpickle.load(f_handle)
-        mid_step = cpickle.load(f_handle)
+    if models is None:
```
Comment (dim) [inline pyAudioAnalysis/audioSegmentation2.py:473]:
```
if type(hmm_model) == str:
  hmm_model = load_hmm_model_file(hmm_model)
```

File: pyAudioAnalysis/audioSegmentation2.py [hunk 4]
```diff
@@ -466,27 +467,25 @@ def save_hmm(hmm_model_name, model, classes, mid_window, mid_step):
...
                                    round(sampling_rate * 0.050),
                                    round(sampling_rate * 0.050))
 
     # apply model
-    labels = hmm.predict(features.T)
+    labels = models.hmm_hmm.predict(features.T)
     labels_gt, class_names_gt, accuracy, cm = \
-        load_ground_truth(gt_file, labels, class_names, mid_step, plot_results)
-    return labels, class_names, accuracy, cm
+        load_ground_truth(gt_file, labels, models.hmm_class_names, models.hmm_mid_step, plot_results)
```
Comment (dim) [inline pyAudioAnalysis/audioSegmentation2.py:487]: I am not sure how efficient that is, sounds like something we should simplify (and keep in memory) too, in a separate PR

File: pyAudioAnalysis/audioTrainTest.py [hunk 5]
```diff
@@ -501,68 +502,49 @@ def feature_extraction_train_regression(folder_name, mid_window, mid_step,
...
-        short_step = cPickle.load(fo)
-        compute_beat = cPickle.load(fo)
-
-    features = np.array(features)
-    labels = np.array(labels)
-    mean = np.array(mean)
-    std = np.array(std)
-
-    classifier = Knn(features, labels, neighbors)
+def load_model_knn(knn_model_name, is_regression=False, models=None):
```
Comment (dim) [inline pyAudioAnalysis/audioTrainTest.py:505]: this should be as simple as:

```
def load_model_knn(file_name, is_regression=False):
  return KNNModel(file_name, is_regression)
```

move everything else into `KNNModel#__init__` (or add a static `KNNModel#load_file` if you prefer) and store e.g. the `self.classifier` there, etc

**Review by dim — CHANGES_REQUESTED** [review]

File: pyAudioAnalysis/modelBundle.py (hunk 1 above)
Comment (mxmCherry) [reply pyAudioAnalysis/modelBundle.py:3]: This was just my initial Friday evening approach. Yes, they can (and better) be separate.

File: pyAudioAnalysis/audioSegmentation2.py (hunk 2 above)
Comment (mxmCherry) [reply pyAudioAnalysis/audioSegmentation2.py:469]: Also keep in mind, that I tried to keep changes minimal and back-compatible.

File: pyAudioAnalysis/audioTrainTest.py (hunk 5 above)
Comment (mxmCherry) [reply pyAudioAnalysis/audioTrainTest.py:505]: Again, @dim I'd prefer not to change much. Maybe even merge the original repo later. Maybe even PR these changes back.

**Pull request closed without merging**